### DISCUSSION TITLE
Timeout with 3 different USB-RS485 adapters

### DISCUSSION THREAD

**Discussion by jardoman** [kellerza/sunsynk#98] [Q&A] [answered]
Hi, 
I have Deye Sun-6kW-sg03lp1-eu inverter and Raspberry Pi 4 with 2GB RAM.  
I tried rs485 to Ethernet/wifi adapter USR-DR404, which is working perfectly with umodbus and also with pymodbus driver(Read_sensors_batch_size need to be set to 8). 
But for some reason I want to use USB to RS485 adapter. I tried 3 USB adapters, but no ones is working correctly. With pymodbus driver they don't work at all. With umodbus driver I have lot of Timeout reading and it is not usable. Cable lenght is 1m, so there shouldn't be a problem. I tried also termination resistor, it didn't helped. It looks like HA or RPI has problem communicating through USB port.

USB adapters, which I tried were:
Wave USB to RS485 Industrial Converter https://www.robotics.org.za/W17286
Basic USB to RS485 Converter, 2 Wire https://www.robotics.org.za/RS485-MINI with FT232RL chip and also with PL2303 chip.

Could somebody help me to solve the problem?
Thank you.

**Comment by kellerza** [comment]
Use mbusd to connect to the ISB and then connect to mbusb from homeassistant

so mbusb performs the same function as your USR

the direct serial driver of both python drivers are not that fault tolerant

**Reply by 197six** [reply, marked as answer]
To expand on this reply, I did the following to make this work after seeing @kellerza's suggestion.

In HA, install "Modbus TCP to Modbus RTU Gateway" Add-on. Configure it to connect to the inverter as per config info (mbusd -d -L - -v 2 -p /dev/ttyUSB0 -s 9600 -m 8N1 -P 502)
Modified the Sunsynk/Deye Inverter Add-on (multi) config to connect via tcp on port: tcp://homeassistant.local:502.

Started Modbus and the inverter add-on and away we go. Seems much more stable (so far) than polling directly over USB to the RS-485 adaptor.

Thanks!
